### ISSUE TITLE
MCS48 RESET

### ISSUE THREAD

**Issue by smdjeff** [mamedev/mame#12916]
I've been investigating some oddities with the Sega G80 Universal Sound board emulation.
It's interesting that the I8035 is allowed to run even while it's being held in reset. Is this an oversight or typical for MAME?
https://github.com/mamedev/mame/blame/324cdb96e05cb09a439b21ff267c713a806ba9dc/src/devices/cpu/mcs48/mcs48.cpp#L1355C2-L1355C2

**Comment by rb6502** [comment]
MAME's scheduler should not run a device that's held in reset.  It's not up to the CPU core itself.

**Comment by smdjeff** [comment]
So this is an issue with segausb.cpp or sega80v.cpp?

**Comment by rb6502** [comment]
What is the issue you're seeing, exactly?

**Comment by smdjeff** [comment]
Put simply, the i8035 continues to retrieve and execute opcode while reset is asserted. It's interesting is that some MCUs specifically do not run op codes when held in reset. So I was thinking that would have been a fix. e.g.
https://github.com/mamedev/mame/blob/master/src/devices/cpu/tms9900/tms9900.cpp#L1276

In the case of the Sega universal sound board, the i8035 runs from RAM loaded by the Z80. The universal sound board uses the msb of it's control line to assert both nRESET and nLOAD. nLOAD is implemented correctly in the segausb.cpp driver.

However, while that's asserted, you can see in the debug logs it's executing garbage at boot and while the RAM is being reloaded by the Z80. This is probably not very noticeable since the factory Z80 code is usually performing many sequential resets and could be mostly hidden by the way the audio analog switches are emulated.  However, since I'm writing Z80 code from scratch for the system, it's easily noticed.

![Screenshot 2024-10-26 at 6 03 18 PM](https://github.com/user-attachments/assets/7463827f-056f-4532-89b3-e92fedae5571)

**Comment by cuavas** [comment]
That doesn’t make sense.  One of the issues I’ve encountered in MAME is that the 4004 and 4040 taking multiple cycles to clear internal dynamic registers on reset can’t be emulated precisely because the execute loop isn’t called while the reset line is asserted.  The framework assumes reset happens instantaneously and the CPU does nothing as long as reset is asserted.

**Comment by happppp** [comment]
segausb.cpp reset line is over here: https://github.com/mamedev/mame/blob/master/src/mame/sega/segausb.cpp#L220
At boot, it makes sure to hold reset: https://github.com/mamedev/mame/blob/master/src/mame/sega/segausb.cpp#L176

Which game are you seeing issues with? I tried Tacscan and it's not misbehaving. If it's a homebrew ROM, make sure to not write 0 to port 0x38 until the i8035 program area is filled.

**Comment by smdjeff** [comment]
I'm testing with StarTrek and it's 0x3F as well. The ROMs quite often writes 0xFF, loads USB_RAM then writes 0x7F.  If you look at the schematic, you'll see that MSB is supposed to reset and assert ram load latch. segausb is implementing it as a single shot, it lets the 8035 run while it's being reloaded. I have no idea why the emulation seems to work well enough anyway.

**Comment by happppp** [comment]
This doesn't apply to Star Trek though? Speech MCU has a ROM, not RAM, so the main CPU doesn't need to stall the MCU.

**Comment by smdjeff** [comment]
Correct, The Universal Sound Board and the Speech board both have 8035s. The USB runs from RAM that's written and most importantly re-written during operation. Not the case for the speech board.

I've made a fork and will propose a patch for review.

**Comment by happppp** [comment]
Are you using the newest and unmodified version of MAME?
I had another look with startrek:

device_reset: MAME holds mcu in reset, MCU is not running.
game writes 0xff to port 0x3f, MCU still stalled
game fills MCU program RAM
game writes 0x7f to port 0x3f, MCU runs
Then it writes 0xff, 0x7f again quickly to reset the MCU

**Comment by happppp** [comment]
At the start:
mcs48_opcode_fetch:0000 15
mcs48_opcode_fetch:0001 27
That is the 2nd MCU of startrek.

If it was the 1st one with the bug you're describing, it would printf:
mcs48_opcode_fetch:0000 00 <-- note the 00
I see nothing like that in your log.
